src/workerd/api/tcp.h

```diff
@@ -0,0 +1,130 @@
+  // stored stream (if the `inner` promise completed).
+  template <typename T>
+  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
+    KJ_IF_MAYBE(e, error) {
+      kj::throwRecoverableException(kj::mv(*e));
+    }
+
+    KJ_IF_MAYBE(s, completedInner) {
+      return f(s);
+    } else {
+      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
+        completedInner = kj::mv(stream);
+        return f(&KJ_ASSERT_NONNULL(completedInner));
+      });
+    }
+  }
+
+  // Stored io stream once the `inner` promise completes.
+  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
+  // Stored error if any of the operations throw.
+  kj::Maybe<kj::Exception> error;
+  // A promise holding a stream that will be available at some point in the future. Typically once
+  // a connection is established.
+  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
+};
+
+struct SocketOptions: public jsg::Object {
+  // TODO(later): TCP socket options will live here.
+  SocketOptions() {};
+public:
+  JSG_RESOURCE_TYPE(SocketOptions, CompatibilityFlags::Reader flags) { }
+};
+
+class Socket: public jsg::Object {
+public:
+  Socket(kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise);
+
+  kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise;
+
+  jsg::Ref<ReadableStream> getReadable() { return KJ_ASSERT_NONNULL(readable).addRef(); }
+  jsg::Ref<WritableStream> getWritable() { return KJ_ASSERT_NONNULL(writable).addRef(); }
+  jsg::Promise<void> getClosed() {
+    // TODO: Right now this promise won't complete when the remote end closes the socket.
+    //       Do we need to wrap ReadableStream/WritableStream to make this work or is there a
+    //       better way?
+    auto& context = IoContext::current();
+    return context.awaitIo(kj::mv(closeFulfiller->promise));
+  }
+
+  jsg::Promise<void> close(jsg::Lock& js);
+  // Closes the socket connection.
+
+  JSG_RESOURCE_TYPE(Socket, CompatibilityFlags::Reader flags) {
+    if (flags.getJsgPropertyOnPrototypeTemplate()) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(readable, getReadable);
+      JSG_READONLY_PROTOTYPE_PROPERTY(writable, getWritable);
+      JSG_READONLY_PROTOTYPE_PROPERTY(closed, getClosed);
+    } else {
+      JSG_READONLY_INSTANCE_PROPERTY(readable, getReadable);
+      JSG_READONLY_INSTANCE_PROPERTY(writable, getWritable);
+      JSG_READONLY_INSTANCE_PROPERTY(closed, getClosed);
+    }
+    JSG_METHOD(close);
+  }
+
+private:
+  kj::Maybe<jsg::Ref<ReadableStream>> readable;
+  kj::Maybe<jsg::Ref<WritableStream>> writable;
+  kj::Promise<kj::Own<kj::AsyncIoStream>> processConnection();
+  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+
+  void visitForGc(jsg::GcVisitor& visitor) {
+    KJ_IF_MAYBE(r, readable) {
+      visitor.visit(*r);
+    }
+    KJ_IF_MAYBE(w, writable) {
+      visitor.visit(*w);
+    }
```

Comment:
Just as a side note... visitor.visit(...) understands maybes and supports multiple arguments, so this could just be `visitor.visit(readable, writable);`